File: Baekjoon/14888_연산자 끼워넣기/B_14888_3.py
import math


def back_track(idx, start_number, addition, subtraction, multiply, division):
    global max_num, min_num
    if idx == N - 1:
        max_num = max(start_number, max_num)
        min_num = min(start_number, min_num)
        return
    else:
        if addition:
            back_track(idx + 1, start_number + numbers[idx + 1], addition - 1, subtraction, multiply, division)

        if subtraction:
            back_track(idx + 1, start_number - numbers[idx + 1], addition, subtraction - 1, multiply, division)

        if multiply:
            back_track(idx + 1, start_number * numbers[idx + 1], addition, subtraction, multiply - 1, division)

        if division:
            if start_number >= 0:
                back_track(idx + 1, start_number // numbers[idx + 1], addition, subtraction, multiply, division - 1)
            else:
                back_track(idx + 1, math.ceil(start_number / numbers[idx + 1]), addition, subtraction, multiply, division - 1)


N = int(input())
numbers = list(map(int, input().split()))
operators = list(map(int, input().split()))

max_num = -float('inf')
min_num = float('inf')

start = numbers[0]
back_track(0, start, operators[0], operators[1], operators[2], operators[3])
print(max_num)
print(min_num)

File: Baekjoon/14888_연산자 끼워넣기/B_14888_3.js
const getMaxAndMinValue = (numbers, operators) => {
  // maxValue 를 음수가 나올 수 있음을 간과하고
  // 처음에 0 으로 했으나 0 으로 하면 안 된다
  let maxValue = -Infinity;
  let minValue = Infinity;
  let cntLimit = N - 1;

  const [plus, minus, multiply, division] = operators

  const getCombinations = (plus, minus, multiply, division, cnt, sums) => {
    if (cnt === cntLimit) {
      // -0 이 나올 수 있음에 주의
      if (sums === -0) {
        sums = 0;
      }

      maxValue = Math.max(maxValue, sums);
      minValue = Math.min(minValue, sums);
      return;
    }

    const targetNumber = numbers[cnt + 1];

    if (plus) {
      getCombinations(plus - 1, minus, multiply, division, cnt + 1, sums + targetNumber);
    }
  
    if (minus) {
      getCombinations(plus, minus - 1, multiply, division, cnt + 1, sums - targetNumber);
    }
  
    if (multiply) {
      getCombinations(plus, minus, multiply - 1, division, cnt + 1, sums * targetNumber);
    }
  
    if (division) {
      getCombinations(plus, minus, multiply, division - 1, cnt + 1, sums < 0 ? Math.ceil(sums / targetNumber) : Math.floor(sums / targetNumber));
    }
  }

  getCombinations(plus, minus, multiply, division, 0, numbers[0]);
  return [maxValue, minValue];
}

const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

// const input = [
//   '2',
//   '5 6',
//   '0 0 1 0'
// ];

// const input = [
//   '3',
//   '3 4 5',
//   '1 0 1 0',
// ];

// const input = [
//   '6',
//   '1 2 3 4 5 6',
//   '2 1 1 1'
// ];

const N = parseInt(input[0]);
const numbers = input[1].split(' ').map(v => parseInt(v));
const operators = input[2].split(' ').map(v => parseInt(v));

const [maxVal, minVal] = getMaxAndMinValue(numbers, operators);
console.log(maxVal);
console.log(minVal);
